### PR TITLE
Tests/setup E2E and Unit tests for Setup

### LINKED ISSUE #276: Add missing e2e and unit tests for `setup` command
Prerequisites
- [ ] #188 
- [ ] #172 
- [ ] #166

Goal: >=90% code coverage.

### PR DESCRIPTION
**Pull request by ctrlc03**
Implemented e2e and unit tests for the Setup action. This PR includes test utilities which can be reused for other tests.

Tests: 

- e2e (full scenarios) -> `e2e/01-setup.test.ts`
- unit:
    * Storage (functions within the `storage.ts` file which include functions used for setting up a ceremony and more) -> `unit/storage.test.ts` 
    * Setup  (helpers functions strictly used for setup)
    
 Renamed `lib.utils.test.ts` to `helpers.utils.test.ts`
 
 Note: the cloud function SetupCeremony now returns the ceremony Id (test env requires reload and clean up)

fix #276

### CODE REVIEW

File: packages/actions/test/e2e/01-setup.test.ts [hunk 1]
```diff
@@ -1,58 +1,225 @@
...
-
-    //     // Remove Auth user.
-    //     await adminAuth.deleteUser(userId)
-
-    //     // Delete admin app.
-    //     await deleteAdminApp()
-    // })
+    // Sample data for running the test.
+    const user = fakeUsersData.fakeUser2
+    const coordinatorEmail = "coordinator@coordinator.com"
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:38]: nit: Make a fake coordinator user.

Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:38]: nit: Make a fake coordinator user.

File: packages/actions/test/e2e/01-setup.test.ts [hunk 2]
```diff
@@ -1,58 +1,225 @@
...
+    const { userApp, userFunctions } = initializeUserServices()
+    const { adminFirestore, adminAuth } = initializeAdminServices()
+    const userAuth = getAuth(userApp)
+
+    // Get configs for storage.
+    const { ceremonyBucketPostfix } = getStorageConfiguration()
+    const ceremonyData = fakeCeremoniesData.fakeCeremonyScheduledFixed
+    const bucketName = getBucketName(ceremonyData.data.prefix, ceremonyBucketPostfix)
+    // file to upload
+    const localPath = "/tmp/circuitMetadata.log"
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:56]: It would appear that you used different paths / filenames conventions (tmp folder, circuitMetadata.log and not circuitprefix_metadata.log, and so on). The purpose is to replicate a ceremony setup following all the conventions and rules.

If the assumption above is true: Why did you change these things?

File: packages/actions/test/e2e/01-setup.test.ts [hunk 3]
```diff
@@ -1,58 +1,225 @@
...
+    })
+
+    it("should fail to create a sample ceremony without being a coordinator", async () => {
+        await signInWithEmailAndPassword(userAuth, user.data.email, userPwd)
+        assert.isRejected(createS3Bucket(userFunctions, bucketName))
+    })
+
+    // run these tests only in production mode
+    if (envType === TestingEnvironment.PRODUCTION) {
+        it("should revert when trying to create a ceremony with an existing prefix", async () => {
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:88]: There is no control other than in the CLI (when ceremony name is requested interactively when running `setup` command) to prevent the duplication of a ceremony. Therefore, we could assume that this test is checking if creating a bucket twice will results in an error. This is one of the mandatory steps to setup a ceremony. On the other hand, there are the Cloud Functions (createCeremony and intEmptyWaitingQueue). 

This would avoid situations where a ceremony could be created on the CF side even without a reserved bucket for artifacts (these two operations are currently nonatomic). Also, this might give us pause and make us consider moving the createS3Bucket functionality directly inside the CF to make the DB write operation + bucket creation atomic. What are your thoughts on this? 

ping to @gurrpi

File: packages/actions/test/e2e/01-setup.test.ts [hunk 4]
```diff
@@ -1,58 +1,225 @@
...
+            // refresh token
+            await currentAuthenticatedCoordinator.getIdToken(true)
+            // Create once
+            await createS3Bucket(userFunctions, duplicateBucketName)
+            await sleep(5000)
+            // Create again
+            expect(createS3Bucket(userFunctions, duplicateBucketName)).to.be.rejectedWith("Failed request.")
+        })
+
+        it("should do a full multi part upload", async () => {
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:101]: It would seem more like a unit test than e2e. What would be the test scenario?

File: packages/actions/test/e2e/01-setup.test.ts [hunk 5]
```diff
@@ -1,58 +1,225 @@
...
+            // 1. create bucket
+            await createS3Bucket(userFunctions, bucketName)
+
+            // 2. multi part upload
+            const success = await multiPartUpload(
+                userFunctions,
+                bucketName,
+                objectName,
+                localPath,
+                process.env.CONFIG_STREAM_CHUNK_SIZE_IN_MB || "128",
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:113]: Avoid using defaults when testing to be sure what values are we using as input for methods.

File: packages/actions/test/e2e/01-setup.test.ts [hunk 6]
```diff
@@ -1,58 +1,225 @@
...
+            await signInWithEmailAndPassword(userAuth, coordinatorEmail, coordinatorPwd)
+
+            const ceremony = fakeCeremoniesData.fakeCeremonyScheduledDynamic
+            const ceremonyBucket = getBucketName(ceremony.data.prefix, `-${randomBytes(5).toString("hex")}`)
+            const circuit = fakeCircuitsData.fakeCircuitSmallNoContributors
+            // 1 create a bucket for the ceremony
+            await createS3Bucket(userFunctions, ceremonyBucket)
+
+            // 2. upload zkey
+            const zkeyLocalFilePath = "/tmp/circuit.zkey"
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:131]: Read comment above

File: packages/actions/test/e2e/01-setup.test.ts [hunk 7]
```diff
@@ -1,58 +1,225 @@
...
+            fs.writeFileSync(zkeyLocalFilePath, "zkey")
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:132]: nit: avoid hard-coded strings for names, paths and fields.

File: packages/actions/test/e2e/01-setup.test.ts [hunk 8]
```diff
@@ -1,58 +1,225 @@
...
+
+            const zkeyStorageFilePath = getZkeyStorageFilePath(circuit.data.prefix!, "circuit_00001.zkey")
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:134]: see nit above

File: packages/actions/test/e2e/01-setup.test.ts [hunk 9]
```diff
@@ -1,58 +1,225 @@
...
+                userFunctions,
+                bucketName,
+                zkeyStorageFilePath,
+                zkeyLocalFilePath,
+                String(process.env.CONFIG_STREAM_CHUNK_SIZE_IN_MB),
+                Number(process.env.CONFIG_PRESIGNED_URL_EXPIRATION_IN_SECONDS)
+            )
+
+            // 3. upload pot
+            const potLocalFilePath = "/tmp/circuit.pot"
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:146]: see nit above

File: packages/actions/test/e2e/01-setup.test.ts [hunk 10]
```diff
@@ -1,58 +1,225 @@
...
+                userFunctions,
+                bucketName,
+                potStorageFilePath,
+                potLocalFilePath,
+                String(process.env.CONFIG_STREAM_CHUNK_SIZE_IN_MB),
+                Number(process.env.CONFIG_PRESIGNED_URL_EXPIRATION_IN_SECONDS)
+            )
+
+            // 4. upload r1cs
+            const r1csLocalFilePath = "/tmp/circuit.r1cs"
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:160]: see nit above

File: packages/actions/test/e2e/01-setup.test.ts [hunk 11]
```diff
@@ -1,58 +1,225 @@
...
+                r1csLocalFilePath,
+                String(process.env.CONFIG_STREAM_CHUNK_SIZE_IN_MB),
+                Number(process.env.CONFIG_PRESIGNED_URL_EXPIRATION_IN_SECONDS)
+            )
+
+            // 5. setup ceremony
+            await setupCeremony(userFunctions, ceremony.data, ceremony.data.prefix!, [circuit.data])
+
+            // 6. confirm
+            const ceremonyDoc = await adminFirestore.collection("ceremonies").doc(ceremony.uid).get()
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:177]: This is definitely not enough. You should check every field value + circuits subcollections documents + S3 bucket paths + files hashes.

File: packages/actions/test/e2e/01-setup.test.ts [hunk 12]
```diff
@@ -1,58 +1,225 @@
...
+            await deleteObjectFromS3(ceremonyBucket, potStorageFilePath)
+            await deleteObjectFromS3(ceremonyBucket, r1csStorageFilePath)
+            await deleteBucket(ceremonyBucket)
+            fs.unlinkSync(zkeyLocalFilePath)
+            fs.unlinkSync(potLocalFilePath)
+            fs.unlinkSync(r1csLocalFilePath)
+        })
+    }
+
+    it("should fail to create a new ceremony when given the wrong path to a zkey", async () => {
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:191]: I'm 99% sure this is a unit test

File: packages/actions/test/e2e/01-setup.test.ts [hunk 13]
```diff
@@ -1,58 +1,225 @@
...
+                objectName,
+                "./nonExistantPath.zkey",
+                process.env.CONFIG_STREAM_CHUNK_SIZE_IN_MB || "128",
+                Number(process.env.CONFIG_PRESIGNED_URL_EXPIRATION_IN_SECONDS) || 3600
+            )
+        )
+    })
+
+    afterAll(async () => {
+        // Clean ceremony and user from DB.
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:209]: The comment says clean ceremony but I cannot see any ceremony doc deletion.

File: packages/actions/test/e2e/01-setup.test.ts [hunk 14]
```diff
@@ -1,58 +1,225 @@
...
+    getPotStorageFilePath,
+    getR1csStorageFilePath,
+    setupCeremony
+} from "../../src"
+import { TestingEnvironment } from "../../src/types/enums"
 
 // Config chai.
 chai.use(chaiAsPromised)
 
 describe("Setup", () => {
```
Comment (0xjei) [inline packages/actions/test/e2e/01-setup.test.ts:35]: It would appear that only the ceremony creation scenario has been tested successfully.

I was wondering, what other scenarios would be worth testing for the V1 release? Perhaps none, but I would put more attention on the output produced at the end of the workflow of the tested scenario (double check the checks and add what's missing, we can't go wrong there).

@gurrpi @daodesigner Do you get any ideas about possible scenarios to test?

File: packages/actions/test/unit/helpers.utils.test.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
+import { expect } from "chai"
+import { fakeCircuitsData } from "../data/samples"
+import { extractPoTFromFilename, extractPrefix, formatZkeyIndex } from "../../src"
+
+describe.skip("Utils", () => {
```
Comment (0xjei) [inline packages/actions/test/unit/helpers.utils.test.ts:5]: Please, be sure to motivate the skip using comments and put a todo to avoid missing these tests in the future.

File: packages/actions/test/unit/services.test.ts [hunk 16]
```diff
@@ -14,9 +14,8 @@ dotenv.config()
  * Unit test for Services helpers.
  * @notice some of these methods are used as a core component for authentication.
  */
-describe("Services", () => {
+describe.skip("Services", () => {
```
Comment (0xjei) [inline packages/actions/test/unit/services.test.ts:17]: Please, be sure to motivate the skip using comments and put a todo to avoid missing these tests in the future.

File: packages/actions/test/unit/setup.test.ts [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+        })
+    })
+    describe("estimatePoT", () => {
+        it("should correctly estimate PoT given the number of constraints", () => {
+            expect(computeSmallestPowersOfTauForCircuit(10e6, 2)).to.be.eq(24)
+        })
+    })
+
+    afterAll(async () => {
+        // Clean ceremony and user from DB.
```
Comment (0xjei) [inline packages/actions/test/unit/setup.test.ts:131]: Did you forget to clean the setupCeremony docs there?

File: packages/actions/test/unit/storage.test.ts [hunk 18]
```diff
@@ -0,0 +1,801 @@
...
+                await signInWithEmailAndPassword(userAuth, coordinatorEmail, coordinatorPwd)
+                // create bucket
+                await createS3Bucket(userFunctions, bucketName)
+                // upload object
+                const success = await multiPartUpload(
+                    userFunctions,
+                    bucketName,
+                    objectName,
+                    localPath,
+                    process.env.CONFIG_STREAM_CHUNK_SIZE_IN_MB || "128",
```
Comment (0xjei) [inline packages/actions/test/unit/storage.test.ts:145]: Same as above

File: packages/actions/test/unit/storage.test.ts [hunk 19]
```diff
@@ -0,0 +1,801 @@
...
+            })
+            it("should fail when providing a non-existent bucket name", async () => {
+                await signInWithEmailAndPassword(userAuth, coordinatorEmail, coordinatorPwd)
+                expect(
+                    multiPartUpload(
+                        userFunctions,
+                        "nonExistentBucketName",
+                        objectName,
+                        localPath,
+                        process.env.CONFIG_STREAM_CHUNK_SIZE_IN_MB || "128",
```
Comment (0xjei) [inline packages/actions/test/unit/storage.test.ts:210]: Same as above

File: packages/actions/test/utils/storage.ts [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+ * Deletes an object from S3 (test function only)
+ * @param bucketName <string> the name of the bucket to delete the object from
+ * @param objectKey <string> the key of the object to delete
+ * @returns <boolean> true if the object was deleted, false otherwise
+ */
+export const deleteObjectFromS3 = async (bucketName: string, objectKey: string) => {
+    if (!process.env.AWS_ACCESS_KEY_ID || !process.env.AWS_SECRET_ACCESS_KEY)
+        // throw new Error("Missing AWS credentials, please add them in the .env file")
+        return false
+    const s3 = new S3Client({
```
Comment (0xjei) [inline packages/actions/test/utils/storage.ts:14]: nit: make an utility to get a new S3Client to avoid code repetition among methods.

File: packages/actions/test/utils/storage.ts [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+/**
+ * Creates mock data on Firestore (test function only)
+ * @param adminFirestore <FirebaseFirestore.Firestore> the admin firestore instance
+ */
+export const createMockCeremony = async (adminFirestore: FirebaseFirestore.Firestore) => {
+    // Create the mock data on Firestore.
+    await adminFirestore
+        .collection(`ceremonies`)
```
Comment (0xjei) [inline packages/actions/test/utils/storage.ts:73]: avoid using hard-coded string, as above

File: packages/actions/test/utils/storage.ts [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+        .doc(fakeCeremoniesData.fakeCeremonyOpenedFixed.uid)
+        .set({
+            ...fakeCeremoniesData.fakeCeremonyOpenedFixed.data
+        })
+
+    await adminFirestore
+        .collection(`ceremonies/${fakeCeremoniesData.fakeCeremonyOpenedFixed.uid}/circuits`)
```
Comment (0xjei) [inline packages/actions/test/utils/storage.ts:80]: avoid hard-coded paths, as above

File: packages/actions/test/utils/storage.ts [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+        })
+}
+
+/**
+ * Cleans up mock data on Firestore (test function only)
+ * @param adminFirestore <FirebaseFirestore.Firestore> the admin firestore instance
+ */
+export const cleanUpMockCeremony = async (adminFirestore: FirebaseFirestore.Firestore) => {
+    await adminFirestore
+        .collection(`ceremonies/${fakeCeremoniesData.fakeCeremonyOpenedFixed.uid}/circuits`)
```
Comment (0xjei) [inline packages/actions/test/utils/storage.ts:93]: same as above

**Review by 0xjei — CHANGES_REQUESTED** [review]
Plenty of new tests! A few inaccuracies and usual recommendations to avoid using hard-coded strings. Left a few comments on e2e setup tests, otherwise just a few nits. Gj :+1: 

I also ask @gurrpi and @daodesigner to take a quick look at the taggs :pray: :heart:

File: packages/actions/test/unit/helpers.utils.test.ts (hunk 15 above)
Comment (ctrlc03) [reply packages/actions/test/unit/helpers.utils.test.ts:5]: My bad here committing this with skip, I only added skip to not run the tests while working on setup

File: packages/actions/test/unit/services.test.ts (hunk 16 above)
Comment (ctrlc03) [reply packages/actions/test/unit/services.test.ts:17]: Same as above, skip was to run only setup tests- will push without skip

File: packages/actions/test/e2e/01-setup.test.ts (hunk 2 above)
Comment (ctrlc03) [reply packages/actions/test/e2e/01-setup.test.ts:56]: Good point, I will amend this

File: packages/actions/test/e2e/01-setup.test.ts (hunk 12 above)
Comment (ctrlc03) [reply packages/actions/test/e2e/01-setup.test.ts:191]: In my mind this was more of: coordinator logs in, tries to setup a ceremony, inputs wrong data (zkey path), setup fail. But I agree that looking like this it's more of a unit test on multiPartUpload working or not.

File: packages/actions/test/e2e/01-setup.test.ts (hunk 3 above)
Comment (ctrlc03) [reply packages/actions/test/e2e/01-setup.test.ts:88]: Tough one here. Where exactly would you move the create S3 bucket function? in the multi part upload? 
Maybe the setup ceremony cloud function should check that a bucket was created and that the required data is inside.  At that point, it would fail if not. 

A frontend check we could do is to check whether the bucket name exists already when getting the ceremony name from the user in the setup phase. This way we can inform them and stop it there (although it will stop later when calling the createS3Bucket function). This can of course be bypassed and would only be true for our "frontend/cli" implementation.

File: packages/actions/test/e2e/01-setup.test.ts (hunk 4 above)
Comment (ctrlc03) [reply packages/actions/test/e2e/01-setup.test.ts:101]: Fair point, I'll move it over.

File: packages/actions/test/e2e/01-setup.test.ts (hunk 5 above)
Comment (ctrlc03) [reply packages/actions/test/e2e/01-setup.test.ts:113]: The same is in the `getStorageConfiguration` test utility that was introduced during the test bootstrapping. Actually forgot it existed (and will use now), but that function uses the same default values if the env var is not set

File: packages/actions/test/e2e/01-setup.test.ts (hunk 14 above)
Comment (daodesigner) [reply packages/actions/test/e2e/01-setup.test.ts:35]: Order of importance for tests is probably: contribution >> finalize > set up; most important thing to test on set up side is probably that resources are locked down as tight as possible: can't override another's ceremony, fails gracefully without affecting other ceremonies (ie. if circom doesn't compile), can't XSS, etc.

Comment (ctrlc03) [reply packages/actions/test/e2e/01-setup.test.ts:35]: We already implemented some good protections on collections data (users) and S3 in general (cannot get pre-signed urls for objects not part of a ceremony -> which are not sensitive) needs a bit more work, although it will also boil down to how the coordinator sets up the IAM role for accessing S3. 
Input validation tests will be added as I get properly down with pentesting this.

File: packages/actions/test/unit/setup.test.ts (hunk 17 above)
Comment (ctrlc03) [reply packages/actions/test/unit/setup.test.ts:131]: Yes, will implement a way to get the ID and delete.

File: packages/actions/test/e2e/01-setup.test.ts (hunk 13 above)
Comment (ctrlc03) [reply packages/actions/test/e2e/01-setup.test.ts:209]: Will be added

File: packages/actions/test/e2e/01-setup.test.ts (hunk 3 above)
Comment (gurrpi) [reply packages/actions/test/e2e/01-setup.test.ts:104]: Looks like the problem here is "How to ensure the uniqueness of the ceremony?". I think rely on S3 bucket name would be a simple approach, since the uniqueness of S3 bucket name is secured by AWS. Then p0tion's responsibility will become only maintain the link securely between ceremony and S3 bucket.

rationale: a ceremony could be considered as a wrapper of S3 bucket (this view will sacrifice agnosticism, but it seems fair for now)

Comment (ctrlc03) [reply packages/actions/test/e2e/01-setup.test.ts:104]: One thing to note is, the process should be stopped once a bucket with the same name is found, as uploading files to the same folder might overwrite if they have the same name.

Comment (gurrpi) [reply packages/actions/test/e2e/01-setup.test.ts:104]: > Maybe the setup ceremony cloud function should check that a bucket was created

Also big plus to this. 
p0tion is kinda simple distributed system (communication between GCP and AWS), so the possibility of request failure, this case is to request to S3, is always present.

If it fails after a few retries over a reasonable time interval, it should probably abort from running the setup command and do a cleanup gracefully

Comment (0xjei) [reply packages/actions/test/e2e/01-setup.test.ts:104]: > Also big plus to this. p0tion is kinda simple distributed system (communication between GCP and AWS), so the possibility of request failure, this case is to request to S3, is always present.

Bravo, you understood what I meant. The CLI (frontend) side controls are there, the backend/CF side controls are missing. I really wanted to emphasize this by saying that that test case was simply half-checking.
 
> If it fails after a few retries over a reasonable time interval, it should probably abort from running the setup command and do a cleanup gracefully

I think testing is also needed for these discussions. BTW, this comment is a great solution.

File: packages/actions/test/e2e/01-setup.test.ts (hunk 5 above)
Comment (0xjei) [reply packages/actions/test/e2e/01-setup.test.ts]: Since we have included an initial check on each individual environment variable, it no longer makes sense to have these defaults.

File: packages/actions/test/e2e/01-setup.test.ts (hunk 14 above)
Comment (0xjei) [reply packages/actions/test/e2e/01-setup.test.ts:44]: > Order of importance for tests is probably: contribution >> finalize > set up; most important thing to test on set up side is probably that resources are locked down as tight as possible: can't override another's ceremony, fails gracefully without affecting other ceremonies (ie. if circom doesn't compile), can't XSS, etc.

Strongly agree w/ the order of importance and thanks @daodesigner for to always pull something interesting out of the hat :heart_decoration: 

> We already implemented some good protections on collections data (users) and S3 in general (cannot get pre-signed urls for objects not part of a ceremony -> which are not sensitive) needs a bit more work, although it will also boil down to how the coordinator sets up the IAM role for accessing S3. Input validation tests will be added as I get properly down with pentesting this.

Thanks @ctrlc03 for the quick and complete answer. Let's add the test case on ceremony override (S3 should avoid creating the same bucket by default and an error should be triggered by CF). Regarding gracefully failure, each ceremony is independent in terms of bucket + collection space and has no relation to other ceremonies spaces. Therefore, if something goes wrong on the ceremony setup side, we just need to make sure to delete the intermediate/incomplete data from the Firestore DB and remove the already created S3 bucket (if any).

File: packages/actions/test/utils/authentication.ts [hunk 24]
```diff
@@ -511,3 +512,64 @@ export const setCustomClaims = async (
...
+ * @param uids <string[]> - the list of uids to delete.
+ */
+export const cleanUpMockUsers = async (
+    adminAuth: Auth,
+    adminFirestore: FirebaseFirestore.Firestore,
+    users: UserDocumentReferenceAndData[]
+): Promise<void> => {
+    for (let i = 0; i < users.length; i++) {
+        await adminAuth.deleteUser(users[i].uid)
+        await adminFirestore.collection("users").doc(users[i].uid).delete()
```
Comment (0xjei) [inline packages/actions/test/utils/authentication.ts:573]: nit: use commonTerms

File: packages/actions/.env.default [hunk 25]
```diff
@@ -37,6 +37,10 @@ CONFIG_CEREMONY_BUCKET_POSTFIX="-ph2-ceremony"
 # The amount of time in seconds which indicates the duration about the validity of a pre-signed URL. 
 # default: 7200 seconds = 2 hours.
 CONFIG_PRESIGNED_URL_EXPIRATION_IN_SECONDS=7200
+# AWS S3 access keys only required for unit/e2e testing (can be ommited)
```
Comment (0xjei) [inline packages/actions/.env.default:40]: nit: missing docs.

File: packages/actions/test/utils/storage.ts [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+export const cleanUpMockCeremony = async (
+    adminFirestore: FirebaseFirestore.Firestore,
+    ceremonyId: string,
+    circuitId: string
+) => {
+    await adminFirestore.collection(getCircuitsCollectionPath(ceremonyId)).doc(circuitId).delete()
+    await adminFirestore.collection(commonTerms.collections.ceremonies.name).doc(ceremonyId).delete()
+}
+
+/// test utils
```
Comment (0xjei) [inline packages/actions/test/utils/storage.ts:124]: These are just for unit testing, right?

**Review by 0xjei — APPROVED** [review]
Few super small nits. Great job, let's merge :)

File: packages/actions/test/utils/storage.ts (hunk 26 above)
Comment (ctrlc03) [reply packages/actions/test/utils/storage.ts:124]: Yes
